File: csla20cs/Templates/EditableChildList.cs
using System;
using System.Collections.Generic;
using System.Text;
using System.Data.SqlClient;
using Csla;

namespace Templates
{
  [Serializable()]
  class EditableChildList : 
    BusinessListBase<EditableChildList, EditableChild>
  {
    #region Factory Methods

    internal static EditableChildList NewEditableChildList()
    {
      return new EditableChildList();
    }

    internal static EditableChildList GetEditableChildList(SqlDataReader dr)
    {
      return new EditableChildList(dr);
    }

    private EditableChildList()
    {
      MarkAsChild();
    }

    private EditableChildList(SqlDataReader dr)
    {
      MarkAsChild();
      Fetch(dr);
    }

    #endregion

    #region Data Access

    private void Fetch(SqlDataReader dr)
    {
      while (dr.Read())
      {
        this.Add(EditableChild.GetEditableChild(dr));
      }
    }


    internal void Update()
    {
      foreach (EditableChild item in DeletedList)
        item.DeleteSelf();
      DeletedList.Clear();

      foreach (EditableChild item in this)
        if (item.IsNew)
          item.Insert();
        else
          item.Update();
    }

    #endregion
  }
}

File: csla20vb/Templates/EditableChildList.vb
Imports System.Data.SqlClient

<Serializable()> _
Public Class EditableChildList
  Inherits BusinessListBase(Of EditableChildList, EditableChild)

#Region " Factory Methods "

  Friend Shared Function NewEditableChildList() As EditableChildList
    Return New EditableChildList
  End Function

  Friend Shared Function GetEditableChildList(ByVal dr As SqlDataReader) As EditableChildList
    Return New EditableChildList(dr)
  End Function

  Private Sub New()
    MarkAsChild()
  End Sub

  Private Sub New(ByVal dr As SqlDataReader)
    MarkAsChild()
    Fetch(dr)
  End Sub

#End Region

#Region " Data Access "

  Private Sub Fetch(ByVal dr As SqlDataReader)

    While dr.Read
      Add(EditableChild.GetEditableChild(dr))
    End While

  End Sub

  Friend Sub Update()

    For Each item As EditableChild In DeletedList
      item.DeleteSelf()
    Next
    DeletedList.Clear()

    For Each item As EditableChild In Me
      If item.IsNew Then
        item.Insert()

      Else
        item.Update()
      End If
    Next

  End Sub

#End Region

End Class
